sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/scripts/run_clm_smddp.py

```diff
@@ -0,0 +1,582 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2020 The HuggingFace Inc. team. All rights reserved.
+# Modifications Copyright 2021 Amazon.com, Inc. or its affiliates. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Fine-tuning the library models for causal language modeling (GPT, GPT-2, CTRL, ...) on a text file or a dataset.
+
+Here is the full list of checkpoints on the hub that can be fine-tuned by this script:
+https://huggingface.co/models?filter=causal-lm
+"""
+# You can also adapt this script on your own causal language modeling task. Pointers for this are left as comments.
+
+import logging
+import math
+import os
+import sys
+from dataclasses import dataclass, field
+from typing import Optional
+
+import datasets
+from datasets import load_dataset
+
+import transformers
+from transformers import (
+    CONFIG_MAPPING,
+    MODEL_FOR_CAUSAL_LM_MAPPING,
+    AutoConfig,
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    HfArgumentParser,
+    Trainer,
+    TrainingArguments,
+    default_data_collator,
+    set_seed,
+)
+from transformers.trainer_utils import get_last_checkpoint
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+# for enable herring
+import smdistributed.dataparallel.torch.torch_smddp
```

Comment:
SMDDP is fully integrated with` transformers`. This should not be necessary.
`transformers` automatically [checks for SMDDP](https://github.com/huggingface/transformers/blob/3e2dd7f92d20f67e34f649cdafbbe28d22a75f44/src/transformers/utils/import_utils.py#L594-L605) and does the [required setup](https://github.com/huggingface/transformers/blob/b79028f0b69886e2a9238482c848aa5556d890b2/src/transformers/training_args.py#L1395-L1396).